### PR TITLE
Dropdown in Windows

### PR DESCRIPTION
**Pull request by sjoshid**
I managed to successfully F up PR #2042. (Patting my back)

This PR adds ability to let dropdowns cross their parent boundary.

SubWindow

### CODE REVIEW

File: druid-shell/src/backend/windows/window.rs
Comment (jneem) [inline druid-shell/src/backend/windows/window.rs:1492]: I think `unwrap_or` here would be simpler than `unwrap_or_else`

**Review by jneem — COMMENTED** [review]
Thanks! Just one very minor suggestion.

File: druid-shell/src/backend/windows/window.rs
Comment (sjoshid) [reply druid-shell/src/backend/windows/window.rs]: Done

File: druid-shell/src/backend/windows/window.rs
Comment (sjoshid) [inline druid-shell/src/backend/windows/window.rs:1889]: @jneem 
I think it would be better to store `WindowLevel` in `WindowHandle`. 
We would then be consistent with what is done in `build()`. ie convert point from window coords to screen coords only when this subwindow is `WindowLevel::Tooltip` or `WindowLevel::Dropdown`. 

What do you think?

Comment (sjoshid) [reply druid-shell/src/backend/windows/window.rs:1889]: There's a `set_level` for `WindowHandle` which is unimpl today. Maybe above suggestion will fix that too.

**Comment by sjoshid** [comment]
I need to fix the changes I made in set_position(). 
Please hold off on merging this.

File: druid-shell/src/backend/windows/window.rs
Comment (jneem) [reply druid-shell/src/backend/windows/window.rs:1889]: Hm, I can see the argument for including `WindowLevel` in `WindowHandle`, but it will require cross-platform changes and I don't think there's a *big* advantage, is there? I mean, you can query the parent easily enough from the windows API (and I think on other platforms too).

Also, I thought we had agreed in #1824 to remove `set_level` from `WindowHandle`.

Comment (sjoshid) [reply druid-shell/src/backend/windows/window.rs:1889]: Few points:
* My point is consistency between `build()` and `set_position()`. Since `build()` calculates positions based on `WindowLevel` so should `set_position`. And not make a platform call which many might not know straight away. (it also negates the need for an extra `unsafe` call if that matters)
* Also there's a bug in `set_position`. I cant use `self.get_position()`. I need parent's position not its own. We can get that if `WindowLevel` (which internally gives us `WindowHandle`) is stored in `WindowHandle`
* To further my previous point, in gtk, parent handle is stored in `WindowHandle` for exactly that purpose. Only difference is _I think_ it's better to store `WindowLevel` (again we get window handle internally) for the sake of first point above. 

_I keep saying store `WindowLevel` in `WindowHandle` but what I actually mean is store it within one of its fields. Like in Windows can store level in `WindowState` which is field of `WindowHandle`_

> Also, I thought we had agreed in #1824 to remove `set_level` from `WindowHandle`.

Ohh. You are right. I can remove it as part of this?

Comment (jneem) [reply druid-shell/src/backend/windows/window.rs:1889]: > I keep saying store `WindowLevel` in `WindowHandle` but what I actually mean is store it within one of its fields.

Ah, got it. I thought you were proposing to put it in the actual cross-platform `WindowHandle` struct, which would then require more invasive changes...

Can you explain the benefit of storing the `WindowLevel` as opposed to just the parent? My understanding of the position calculation was that it should only depend on whether there's a parent or not.

Comment (jneem) [reply druid-shell/src/backend/windows/window.rs:1889]: > Ohh. You are right. I can remove it as part of this?

Please!

Comment (sjoshid) [reply druid-shell/src/backend/windows/window.rs:1889]: > Can you explain the benefit of storing the WindowLevel as opposed to just the parent?

So instead of doing this:
```
if let Some(p) = self.parent {
    // use p to calculate position
}
```

We can do this:
```
if WindowLevel::Tooltip(p) || WindowLevel::Dropdown(p) {
    // use p to calculate position
}
```

`WindowLevel` check is consistent with `build()`

Comment (jneem) [reply druid-shell/src/backend/windows/window.rs:1889]: ...but I think you also want to include `WindowLevel::Modal(p)` there, because the only difference between those 3 levels is their z-order. They should all treat positioning in the same way.

Comment (sjoshid) [reply druid-shell/src/backend/windows/window.rs:1889]: True. But at least in Windows it is unimplemented/unfinished. I didnt include it because I that change should be included when modal feature is implemented.

Comment (sjoshid) [reply druid-shell/src/backend/windows/window.rs:1889]: Not sure if you agree to my previous point but Im going to resolve this conversation. It's been a while. If you disagree, please let me know and I'll add `WindowLevel::Modal(p)`.

Comment (jneem) [reply druid-shell/src/backend/windows/window.rs:1889]: Oh, sorry, I had forgotten about this, but I don't understand the part about modal not being implemented in windows. The only difference between (druid terminology) tooltip, modal, and dropdown is that they have different z-order. They should have exactly the same behavior in all other respects, and so there's no reason to support, say, dropdown without also supporting modal.

Comment (sjoshid) [reply druid-shell/src/backend/windows/window.rs:1889]: Where is the best place in code to see different z-orders in action? In Windows at least, z-order is same for tooltips and dropdown (and maybe for modal) ie 1 + max of existing z-orders.

Comment (jneem) [reply druid-shell/src/backend/windows/window.rs:1889]: It seems like the mac backend is the only one that handles `WindowLevel` according to the documentation (see `as_raw_window_level`), but I don't have a mac so I can't test it. The gtk backend does something different. I guess what happened was that `WindowLevel` was created according to the mac behavior, and then because the variants have suggestive-sounding names it was implemented differently on gtk (and partially on windows).

But I think the right way forward is the following: let's first make window parent/child positioning work. After that, we can worry about z-order and any other behavioral changes that different subwindows can have. We'll need to figure out what we can reliably support cross-platform.

Comment (sjoshid) [reply druid-shell/src/backend/windows/window.rs]: Sounds good!!
I have made changes as per this conversation.

**Comment by sjoshid** [comment]
^^
This is ready to merge (assuming there are no more code review changes)

File: druid-shell/src/backend/windows/window.rs
Comment (jneem) [inline druid-shell/src/backend/windows/window.rs:240]: This never changes, right? So you don't need the `RefCell`. And I think `AppWindow` is a good default so we don't need the `Option` either.

File: druid-shell/src/backend/windows/window.rs
Comment (jneem) [inline druid-shell/src/backend/windows/window.rs:1346]: Can you add `Modal` here also?

File: druid-shell/src/backend/windows/window.rs
Comment (jneem) [inline druid-shell/src/backend/windows/window.rs:1405]: Maybe `Modal` here too? In principle, this "regresses" the topmost behavior of modal, but I'm pretty sure the current code is dead code anyway, because `set_level` doesn't currently set modal.

File: druid-shell/src/backend/windows/window.rs
Comment (jneem) [inline druid-shell/src/backend/windows/window.rs:1903]: Ah ok, I see what the `Option<WindowLevel>` is doing now. I think I prefer getting rid of the option still, because then the types will get rid of this dead code.

File: druid-shell/src/backend/windows/window.rs
Comment (jneem) [inline druid-shell/src/backend/windows/window.rs:1910]: I don't think there's a need for this else. If the state doesn't upgrade, it should mean that the window was closed.

File: druid-shell/src/backend/windows/window.rs
Comment (sjoshid) [reply druid-shell/src/backend/windows/window.rs]: Dang! Forgot to revert this after removing `set_level`. Thanks.

File: druid-shell/src/backend/windows/window.rs
Comment (sjoshid) [reply druid-shell/src/backend/windows/window.rs]: Added `Modal` and removed the `match` statement.

File: druid-shell/src/backend/windows/window.rs
Comment (sjoshid) [reply druid-shell/src/backend/windows/window.rs]: Default as `AppWindow` made this easier. Fixed this and your suggestion below. Thanks.

File: druid-shell/src/backend/windows/window.rs
Comment (sjoshid) [reply druid-shell/src/backend/windows/window.rs]: Added `Modal`

**Review by jneem — APPROVED** [review]
Thanks!

**Comment by sjoshid** [comment]
^ Thank you too. 
Im going to pick your brain one last time.
Recently installed linux to understand more about gtk because it always comes up but Im seeing.........bizarre behavior to say the least. 
My assumption was subwindow positioning was impl recently in gtk meaning it was working.
Is the assumption wrong?

https://user-images.githubusercontent.com/6751090/147617220-d240169c-d9e6-469b-9d28-f5981dbb042b.mp4

https://user-images.githubusercontent.com/6751090/147617185-547673eb-81d4-4e63-834e-b2b1d150c52e.mp4

**Comment by Maan2003** [comment]
@sjoshid can you check the x11 backend (`cargo run --example subwindow --features x11 --no-default-features`)

**Comment by jneem** [comment]
Thanks, I'll look into those. It's highly likely that some things didn't get properly updated when we changed to relative positions.

**Comment by sjoshid** [comment]
@Maan2003 I cant. :( Im running into build errors one after another. Sorry. 

@jneem Thanks.

**Comment by sjoshid** [comment]
I know Im late....but I fixed build errors (like `libclang` not found, etc.) for x11. 
Happy to report x11 is working as expected (to Windows). The tooltip position is off in `sub_window` example but other than that it looks good.

**Comment by Maan2003** [comment]
nice :). please create an issue about tooltip positioning with the desktop enviroment version, screenshots, ... I tried fixing positioning but it is hard to fix for all enviroments on linux.

**Comment by sjoshid** [comment]
Done. #2100

**Comment by jneem** [comment]
@Maan2003 I think I found a little bug with DP scaling and get_position (#2101).

**Pull request merged**
